vumi/transports/turn/turn.py

```diff
@@ -0,0 +1,205 @@
+            request.write(msg)
+        except ValueError, e:
+            request.setResponseCode(http.BAD_REQUEST)
+            msg = "ValueError: %s" % e
+            log.msg('Returning %s: %s' % (http.BAD_REQUEST, msg))
+            request.write(msg)
+        except Exception, e:
+            request.setResponseCode(http.INTERNAL_SERVER_ERROR)
+            log.err("Error processing request: %s" % (request,))
+        request.finish()
+
+    def render(self, request):
+        self.do_render(request)
+        return NOT_DONE_YET
+
+
+class HttpResponseHandler(Protocol):
+    def __init__(self, deferred):
+        self.deferred = deferred
+        self.stringio = StringIO()
+
+    def dataReceived(self, bytes):
+        self.stringio.write(bytes)
+
+    def connectionLost(self, reason):
+        self.deferred.callback(self.stringio.getvalue())
+
+
+class TurnTransport(Transport):
+
+    agent_factory = None  # For swapping out the Agent we use in tests.
+
+    def mkres(self, cls):
+        resource = cls(self.config, self.publish_delivery_report, self.publish_message)
+        self._resources.append(resource)
+        return (resource, self.config['web_%s_path' % (path_key,)])
+
+    @inlineCallbacks
+    def setup_transport(self):
+        self._resources = []
+        resources = [self.mkres(ReceiveEventResource)]
+        self.receipt_resource = yield self.start_web_resources(
+            resources, self.config['web_port'], LogFilterSite)
+
+    def get_transport_url(self):
+        """
+        Get the URL for the HTTP resource. Requires the worker to be started.
+
+        This is mostly useful in tests, and probably shouldn't be used in
+        non-test code, because the API might live behind a load balancer or
+        proxy.
+        """
+        addr = self.receipt_resource.getHost()
+        return "http://%s:%s" % (addr.host, addr.port)
+
+    @inlineCallbacks
+    def handle_outbound_message(self, message):
+        """
+        handle messages arriving over AMQP meant for delivery via turn
+        """
+
+        message_params = {
+            "preview_url": False,
+            "recipient_type": "individual",
+            "to": normalize_outbound_msisdn(message['to_addr']),
+            "type": "text",
+            "text": {"body": message['content']}
+        }
+
+        log.msg('Hitting %s with %s' % (self.config['url'], message_params))
+        log.msg(urlencode(message_params))
+
+        try:
+            response = yield http_request_full(
+                self.config['url'], message_params.to_json(), {
+                    'User-Agent': ['Vumi Turn Transport'],
+                    'Content-Type': ['application/json'],
+                    'Authorization': ['Bearer {}'.format(TURN_TOKEN)],
+                    }, 'POST', agent_class=self.agent_factory)
+        except ConnectionRefusedError:
+            log.msg("Connection failed sending message:", message)
+            raise TemporaryFailure('connection refused')
```

Comment:
Instead of using this old vumi method (it has it's issues), I think we should rather rely on `treq`